indimail-x/proxyimap.8
.TH proxyimap 8
.SH NAME
.PP
proxyimap \- IMAP Proxy
.PP
proxypop3 \- POP3 Proxy
.PP
vproxy \- Generic Proxy

.SH SYNOPSYS
.PP
\fBproxyimap\fR \fIpath_of_imapd\fR \fIMaildir\fR
.PP
\fBproxypop3\fR \fIpath_of_pop3d\fR \fIMaildir\fR
.PP
\fBvproxy\fR \fIhost\fR \fIport\fR [\fIlogin_sequence\fR]

.SH DESCRIPTION
\fBproxyimap\fR and \fBproxypop3\fR act as proxy for IMAP4 and POP3 protocol in a clustered
domain where users homedir lie on multiple hosts. \fBproxyimap\fR and \fBproxypop3\fR correctly
locates the user's mailstore and make a IMAP4 or POP3 connection to the user's mailstore host.
This facilitates to have a single IP address or DNS entry for all hosts comprising a clustered
domain. The proxy works by passing all data from the file descriptor 0 to the socket connection
(IMAP4 or POP3). All data which comes from the socket is is passed to file descriptor 1.
The recommended method for using \fBproxyimap\fR and \fBproxypop3\fR is through tcpserver(1).
It is recommended to use a cdb files (-x option in tcpserver) tcp.imap for \fBproxyimap\fR and
tcp.pop3 for \fBproxypop3\fR. For enabling unrestricted IMAP from a host, set environment
variable POP3CLIENT and IMAPCLIENT in cdb file used by the respective tcpservers (compiled
by tcprules). e.g. To allow unrestricted IMAP/POP3 access from the host with IP w.x.y.z, use
tcprules to compile the following.

.EX
tcp.imap
w.x.y.z:allow,IMAPCLIENT=""

tcp.pop3
w.x.y.z:allow,POP3CLIENT=""
.EE

\fBproxyimap\fR and \fBproxypop3\fR supports TLS/SSL and STARTTLS sessions.
For TLS/SSL session you can use \fBtcpserver\fR(1) with -n option or use
the \fBcouriertls\fR(1) program.

If you set the \fBAUTOADDUSER\fR environment variable, users can be automatically added by
using the \fBadminclient(8)\fR protocol to connect to \fBindisrvr(8)\fR. The auto provisioning
happens only if the user does not exist in IndiMail's authentication database. If you set the
\fBADDUSERCMD\fR environment variable to a program, the program will be run with the password
passed as an argument.

Following environment variables are needed by these two proxies to start up

.TP 4
\fBTCPREMOTEIP\fR
IP address of remote host.
.TP
\fBTCPLOCALPORT\fR
Local TCP/IP Port on which proxyimap, proxypop3 is running.
.TP
\fBDESTPORT\fR
Of the form protocol:port where protocol is either imap or pop3. Port is the TCP/IP port on
the remote host offering the imap or pop3 protocol.
.TP
\fBCOURIERTLS\fR
The program to run to enable TLS/SSL session if the client issues the
\fBSTARTTLS\fR command. If not set, the program \fBsslerator\fR(8) will be
used. One can also use dotls(1) or the couriertls(1) programs. In fact, you
can use any program that can encrypt data on descriptors 0 and 1 and can
print a BANNER on descriptor 1 before initiating a TLS/SSL session. This
banner should be one of the following

.EX
for imap
IMAPTAG Begin SSL/TLS negotiation now.

for pop3
+OK Begin SSL/TLS negotiation now.
.EE

.PP
If you use the AUTOADDUSER environemnt variable to provision new users,
adminclient will be used to execute various indimail commands. In this case
additional environment variables will needed
.TP 4
\fBADMIN_USER\fR
An existing user set using mgmtpass(8) program.
.TP
\fBADMIN_PASS\fR
Password for the user
.TP
\fBADMIN_HOST\fR
Host running indisrvr(8)
.TP
\fBADMIN_PORT\fR
port on which indisrvr(8) is running
.TP
\fBCERTFILE\fR
openssl certificate to use in case indisrvr(8) is running in TLS mode
.TP
\fBCAFILE\fR
openssl cafile in case CA certificates is needed.
.TP
\fBMATCH_CN
Match host with Common Name of the certificate that indisrvr(8) is running

.PP
Currently proxyimap and proxypop3 allow 5 bad login attempts, after which the connection is
terminated.

.PP
It is also possible to set data timeout for the proxies through the environment variable
DATA_TIMEOUT. The two proxies will terminate if it finds no data coming from either the
client or the actual imap/pop3 server within the time in seconds specified by DATA_TIMEOUT.

.PP
vproxy is a generic proxy which passes data from file descriptor 0 to a TCP/IP connection and
data from the TCP/IP connection to file descriptor 1. The TCP/IP connection is specified by
\fIhost\fR and \fIport\fR parameters. login_sequence specifies initial data (if needed), that
needs to be supplied to the application listening on port port on host host.

.SH OPTIONS
.TP 4
\fIpath_of_imapd\fR
Specifies the full path of IndiMail's imapd executable
.TP
\fIpath_of_pop3d\fR
Specifies the full path of IndiMail's pop3d executable
.TP
\fIhost\fR
Host Name (Host Name or IP address for which the proxy service is needed.
.TP
\fIport\fR
TCP/IP Port on host for which proxy services is needed.
.TP
\fIlogin_sequence\fR
Initial data needed by some servers to start up properly. e.g. POP3 service will require USER
username and PASS passwd sequence to login.

.SH RETURN VALUE
0 for success and 1 for failure in which case all diagnostics are printed on stdout.

.SH "SEE ALSO"
authindi(8), adminclient(8), indisrvr(8), sslerator(8), dotls(1),
tcpserver(1), couriertls(1).


indimail-x/NEWS
doc/RELEASE-Notes-indimail.md

indimail-x/proxyimap.c
/*
 * $Log: proxyimap.c,v $
 * Revision 1.3  2022-12-25 20:32:53+05:30  Cprogrammer
 * allow any TLS/SSL helper program other than sslerator, couriertls
 *
 * Revision 1.2  2019-04-22 23:14:34+05:30  Cprogrammer
 * added missing strerr.h
 *
 * Revision 1.1  2019-04-18 08:39:50+05:30  Cprogrammer
 * Initial revision
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifndef	lint
static char     sccsid[] = "$Id: proxyimap.c,v 1.3 2022-12-25 20:32:53+05:30 Cprogrammer Exp mbhangui $";
#endif

#ifdef CLUSTERED_SITE
#ifdef HAVE_CTYPE_H
#include <ctype.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_SIGNAL_H
#include <signal.h>
#endif
#ifdef HAVE_QMAIL
#include <stralloc.h>
#include <strerr.h>
#include <strmsg.h>
#include <env.h>
#include <str.h>
#include <fmt.h>
#include <scan.h>
#include <subfd.h>
#include <getln.h>
#include <substdio.h>
#endif
#include "AuthModuser.h"
#include "proxylogin.h"

static void     bye();
static char     strnum[FMT_ULONG + 11];
static stralloc tmpbuf = {0};

static void
die_nomem()
{
	strerr_warn1("proxyimap: out of memory", 0);
	_exit(111);
}

int
main(int argc, char **argv)
{
	static stralloc username = {0}, password = {0}, imaptag = {0}, dummy2 = {0}, line = {0};
	int             c, ret, match;
	char           *ptr, *cptr, *remoteip, *local_port, *destport;
	char           *tag = env_get("IMAPLOGINTAG");
	char           *binqqargs[7];

	if (argc != 3) {
		strerr_warn1("usage: proxyimap imapd Maildir", 0);
		return(1);
	}
	remoteip = local_port = destport = (char *) 0;
	if (!(remoteip = env_get("TCPREMOTEIP"))) {
		strerr_warn1("ERR: TCPREMOTEIP not set", 0);
		_exit(1);
	} else
	if (!(local_port = env_get("TCPLOCALPORT"))) {
		strerr_warn1("ERR: TCPLOCALPORT not set", 0);
		_exit(1);
	} else
	if (!(destport = env_get("DESTPORT"))) {
		strerr_warn1("ERR: DESTPORT not set", 0);
		_exit(1);
	} 
	signal(SIGALRM, bye);
	if (AuthModuser(argc, argv, 60, 5)) {
		strmsg_out1("* OK IMAP4rev1 Server Ready.\r\n");
		strerr_warn3("INFO: Connection, remoteip=[", remoteip, "]", 0);
		if (!env_put2("BADLOGINS", "0"))
			die_nomem();
	} else {
		if (!(ptr = (char *) env_get("BADLOGINS"))) {
			c = 0;
			if (!env_put2("BADLOGINS", "1"))
				die_nomem();
		} else {
			scan_uint(ptr, (unsigned int *) &c);
			strnum[fmt_uint(strnum, (unsigned int) (c + 1))] = 0;
			if (!env_put2("BADLOGINS", strnum))
				die_nomem();
		}
		if (c > 3)
			_exit(1);
		strmsg_out2(tag ? tag : "", tag ? " NO Login failed.\r\n" : "NO Login failed.\r\n");
		strerr_warn5("ERR: LOGIN FAILED, remoteip=[", remoteip, "] badlogin=[", strnum, "]", 0);
	}
	alarm(0);
	for (;;) {
		alarm(60);
		if (getln(subfdinsmall, &line, &match, '\n') == -1) {
			strerr_warn1("proxyimap: read-stdin: ", &strerr_sys);
			return (-1);
		}
		if (!match || !line.len)
			break;
		if (match) {
			line.len--;
			line.s[line.len] = 0;
		} else {
			if (!stralloc_0(&line))
				die_nomem();
			line.len--;
		}
		alarm(0);
		ret = 0;
		/*- imaptag */
		for (ptr = line.s; *ptr && isspace(*ptr); ptr++);
		for (c = 0, cptr = ptr; *ptr && !isspace(*ptr); c++, ptr++);
		if (c) {
			if (!stralloc_copyb(&imaptag, cptr, c) ||
					!stralloc_0(&imaptag))
				die_nomem();
			imaptag.len--;
			ret++;
		}

		/*- dummy2 */
		for (; *ptr && isspace(*ptr); ptr++);
		for (c = 0, cptr = ptr; *ptr && !isspace(*ptr); c++, ptr++);
		if (c) {
			if (!stralloc_copyb(&dummy2, cptr, c) ||
					!stralloc_0(&dummy2))
				die_nomem();
			dummy2.len--;
			ret++;
		}

		/*- username */
		for (; *ptr && isspace(*ptr); ptr++);
		for (c = 0, cptr = ptr; *ptr && !isspace(*ptr); c++, ptr++);
		if (c) {
			if (!stralloc_copyb(&username, cptr, c) ||
					!stralloc_0(&username))
				die_nomem();
			username.len--;
			ret++;
		}

		/*- password */
		for (; *ptr && isspace(*ptr); ptr++);
		for (c = 0, cptr = ptr; *ptr && !isspace(*ptr); c++, ptr++);
		if (c) {
			if (!stralloc_copyb(&password, cptr, c) ||
					!stralloc_0(&password))
				die_nomem();
			password.len--;
			ret++;
		}

		if (ret >= 2) {
			for (ptr = dummy2.s;*ptr;ptr++) {
				if (islower((int) *ptr))
					*ptr = toupper((int) *ptr);
			}
			if (!str_diff(dummy2.s, "LOGIN")) /*- proxylogin should normally never return */
				return(proxylogin(argv, destport, username.s, password.s, remoteip, imaptag.s, 1));
			else
			if (!str_diff(dummy2.s, "NOOP")) {
				strmsg_out2(imaptag.s, " OK NOOP completed\r\n");
				continue;
			} else
			if (!str_diff(dummy2.s, "STARTTLS")) {
				char           *p;

				if (!env_put2("AUTHARGC", "0"))
					die_nomem();
				for (c = 0; c < argc; c++) {
					p = strnum;
					p += fmt_strn(p, "AUTHARGV", 8);
					p += fmt_uint(p, (unsigned int) c);
					*p++ = 0;
					if (!env_unset(strnum))
						die_nomem();
				}
				if (!env_unset("BADLOGINS") || !env_unset("IMAP_STARTTLS"))
					die_nomem();
				alarm(0);
				if (!(ptr = env_get("COURIERTLS")))
					ptr = PREFIX"/bin/sslerator";
				c = str_rchr(ptr, '/');
				p = ptr[c] ? ptr + c + 1 : ptr;
				if (str_diff(p, "couriertls")) {
					binqqargs[0] = ptr;
					binqqargs[1] = argv[0];
					binqqargs[2] = argv[1]; 
					binqqargs[3] = argv[2];
					binqqargs[4] = 0;
					if (!stralloc_copy(&tmpbuf, &imaptag) ||
							!stralloc_catb(&tmpbuf, " Begin SSL/TLS negotiation now.\r\n", 33) ||
							!stralloc_0(&tmpbuf))
						die_nomem();
					if (!env_put2("BANNER", tmpbuf.s))
						die_nomem();
				} else {
					if (!env_unset("COURIERTLS"))
						die_nomem();
					binqqargs[0] = ptr;
					binqqargs[1] = "-remotefd=0";
					binqqargs[2] = "-server";
					binqqargs[3] = argv[0];
					binqqargs[4] = argv[1]; 
					binqqargs[5] = argv[2];
					binqqargs[6] = 0;
					strmsg_out2(imaptag.s, " Begin SSL/TLS negotiation now.\r\n");
				}
				execv(*binqqargs, binqqargs);
				strerr_warn3("proxyimap: execv: ", *binqqargs, ": ", &strerr_sys);
				continue;
			} else
			if (!str_diff(dummy2.s, "CAPABILITY")) {
				imapd_capability();
				strmsg_out2(imaptag.s, " OK CAPABILITY completed\r\n");
				continue;
			} else
			if (!str_diff(dummy2.s, "LOGOUT")) {
				strmsg_out1("* BYE IMAP4rev1 server shutting down\r\n");
				strmsg_out1(" OK LOGOUT completed\r\n");
				break;
			}
		}
		strmsg_out1("* NO Error in IMAP command received by server.\n");
	}
	return(0);
}

static void bye()
{
	strerr_warn1("ERR: TIMEOUT", 0);
	_exit(0);
}
#else
#include <strerr.h>
int
main()
{
	strerr_warn1("IndiMail not configured with --enable-user-cluster=y", 0);
	return(0);
}
#endif
